include/gz/utils/Subprocess.hh

```diff
@@ -32,46 +36,96 @@ namespace gz
 namespace utils
 {
 
+/// \brief Create a RAII-type object that encapsulates a subprocess.
 class Subprocess
 {
+  /// \brief Constructor
+  ///
+  /// This variant will spawn a subprocess that inherits the environment
+  /// from the calling process.
+  ///
+  /// \param[in] _commandLine set of arguments starting with an executable
+  ///   used to spawn the subprocess
+  public: explicit Subprocess(const std::vector<std::string> &_commandLine):
+    commandLine(_commandLine),
+    environment({}),
+    inheritEnvironment(true)
+  {
+    this->Create();
+  }
+
+  /// \brief Constructor
+  ///
+  /// This variant will spawn a subprocess that uses the user-specified
+  /// environment
+  ///
+  /// \param[in] _commandLine set of arguments starting with an executable
+  ///   used to spawn the subprocess
+  /// \param[in] _environment environment variables to set in the spawned
+  ///   subprocess
   public: Subprocess(const std::vector<std::string> &_commandLine,
-                     const std::vector<std::string> &_environment = {}):
+                     gz::utils::EnvironmentMap _environment):
     commandLine(_commandLine),
-    environment(_environment)
+    environment(std::move(_environment)),
+    inheritEnvironment(false)
   {
     this->Create();
   }
 
+  /// \brief Constructor
+  ///
+  /// This variant will spawn a subprocess that uses the user-specified
+  /// environment
+  ///
+  /// \param[in] _commandLine set of arguments starting with an executable
+  ///   used to spawn the subprocess
+  /// \param[in] _environment environment variables to set in the spawned
+  ///   subprocess
+  public: Subprocess(const std::vector<std::string> &_commandLine,
+                     const gz::utils::EnvironmentStrings &_environment):
+    Subprocess(_commandLine, gz::utils::envStringsToMap(_environment))
+  {
+  }
+
+
   private: void Create()
   {
-    if (this->process)
+    if (this->process != nullptr)
         return;
 
     this->process = new subprocess_s;
 
+    std::vector<std::string> environmentStr;
+    std::vector<const char*> environmentCstr;
     std::vector<const char*> commandLineCstr;
+
     for (const auto &val : this->commandLine)
     {
       commandLineCstr.push_back(val.c_str());
     }
     commandLineCstr.push_back(nullptr);
 
-    std::vector<const char*> environmentCstr;
-    for (const auto &val : this->environment)
+    if (!this->inheritEnvironment)
     {
-      environmentCstr.push_back(val.c_str());
+      for (const auto &[key, val] : this->environment)
```

Comment:
nit: This can use the new map -> vector function.

Reply:
Good call, done in a81031c